File: Codigos_1_de_2/Ordenamiento.py
def main():
    #Conjunto de datos inicial arary de edades_python
    edades_python = [25, 30, 42, 19, 20, 55, 33, 28, 36, 40, 22, 27]
    #crearemos un pequeño menu par aejeutar los algoritomos de Ordenamiento
    while True:
        # Mostramos el menú
        print("\nMenú para ejecutar los algoritmos de ordenamiento a nuestro Array: edades_python = [25, 30, 42, 19, 20, 55, 33, 28, 36, 40, 22, 27]" )
        print("1. Insertion Sort (Ordenamiento por inserción)")
        print("2. Bubble Sort (Ordenamiento de burbuja)")
        print("3. Selection Sort (Ordenamiento por selección)")
        print("4. Merge Sort (Ordenamiento por mezcla)")
        print("5. Random Sort (Ordenamiento aleatorio)")
        print("6. Salir")

        #pedimos ingrese una opcion por teclado
        opcion = input("Ingresa un número del 1 al 5 para ejecutar un algoritmo de ordenamiento (o 6 para salir): ")

        # Evaluamos la opción ingresada
        if opcion == '1':
            O_Insercion(edades_python)
            print("Array ordenado por inserción:", edades_python)
        elif opcion == '2':
            O_Burbuja(edades_python)
            print("Array ordenado por burbuja:", edades_python)
        elif opcion == '3':
            O_Select(edades_python)
            print("Array ordenado por selección:", edades_python)
        elif opcion == '4':
            O_Mezcla(edades_python)
            print("Array ordenado por mezcla:", edades_python)
        elif opcion == '5':
            O_Random(edades_python)
            print("Array ordenado aleatoriamente:", edades_python)
        elif opcion == '6':
            print("Saliendo del programa...")
            break
        else:
            #cualqueir otro digito pro fuera de los elegidos incialmente se notifica quee sta incorrecto
            print("Opción inválida. Por favor, ingresa un número del 1 al 6.")

# empezamos a crear todos los metodos para cad uno de los algoritmos de ordenamiento

def O_Insercion(edadesi):
    # Iteramos sobre cada elemento del array
    for i in range(1, len(edadesi)):
        # Guardamos el valor actual para comparar y colocar en la posición correcta
        current_value = edadesi[i]
        # Inicializamos el índice para comparar con el valor actual
        j = i - 1
        # Movemos los elementos mayores que el valor actual una posición hacia adelante
        while j >= 0 and edadesi[j] > current_value:
            edadesi[j + 1] = edadesi[j]
            j -= 1
        # Colocamos el valor actual en la posición correcta
        edadesi[j + 1] = current_value
    return edadesi


def O_Burbuja(edadesb):
    n = len(edadesb)
    # Iteramos sobre todos los elementos del array
    for i in range(n):
        # En cada iteración, el elemento más grande se moverá hacia el final,
        # por lo que podemos reducir el rango de comparación
        for j in range(0, n - i - 1):
            # Comparamos elementos adyacentes y los intercambiamos si están en el orden incorrecto
            if edadesb[j] > edadesb[j + 1]:
                edadesb[j], edadesb[j + 1] = edadesb[j + 1], edadesb[j]
    return edadesb

def O_Select(edadess):
    # Iteramos sobre cada elemento del array
    for i in range(len(edadess)):
        # Encontramos el índice del elemento mínimo en el subarray no ordenado
        indice_minimo = i
        for j in range(i + 1, len(edadess)):
            if edadess[j] < edadess[indice_minimo]:
                indice_minimo = j
        # Intercambiamos el elemento mínimo encontrado con el primer elemento del subarray no ordenado
        edadess[i], edadess[indice_minimo] = edadess[indice_minimo], edadess[i]
    return edadess


def O_Mezcla(edadesm):
    # Si el tamaño del array es mayor que 1, se necesita ordenar de lo contrario no
    if len(edadesm) > 1:
        # Encontrar el punto medio del array
        mid = len(edadesm) // 2
        
        # Dividir el array en dos mitades
        left_half = edadesm[:mid]
        right_half = edadesm[mid:]

        # Llamada recursiva para ordenar ambas mitades
        O_Mezcla(left_half)
        O_Mezcla(right_half)

        # Fusionar las mitades ordenadas
        i = j = k = 0
        
        # Iterar sobre las dos mitades y fusionarlas
        while i < len(left_half) and j < len(right_half):
            # Comparar los elementos de las dos mitades y agregar el más pequeño al array resultante
            if left_half[i] < right_half[j]:
                edadesm[k] = left_half[i]
                i += 1
            else:
                edadesm[k] = right_half[j]
                j += 1
            k += 1

        # Añadir los elementos restantes de la mitad izquierda, si hay alguno
        while i < len(left_half):
            edadesm[k] = left_half[i]
            i += 1
            k += 1

        # Añadir los elementos restantes de la mitad derecha, si hay alguno
        while j < len(right_half):
            edadesm[k] = right_half[j]
            j += 1
            k += 1
    return edadesm

import random

def O_Random(edadesr):
    n = len(edadesr)
    # Realizamos un número suficiente de intercambios aleatorios
    # para mezclar el array
    for _ in range(n * n):
        # Elegimos dos índices aleatorios en el arreglo
        i, j = random.randint(0, n - 1), random.randint(0, n - 1)
        # Intercambiamos los elementos en esos índices
        edadesr[i], edadesr[j] = edadesr[j], edadesr[i]
    return edadesr


if __name__ == "__main__":
    main()

File: Codigos_1_de_2/Ordenamiento.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Project/Maven2/JavaApp/src/main/java/${packagePath}/${mainClassName}.java to edit this template
 */
package co.edu.udc.ordenamiento_java;

import java.util.Random;
import java.util.Scanner;

/**
 *
 * @author PEDRO DE HORTA
 */
public class Ordenamiento {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        //definimos el arreglo edades_java
        int[] edades_java = {25, 30, 42, 19, 20, 55, 33, 28, 36, 40, 22, 27};

        //Cantidad de ventas
        System.out.println("\n--------------------Bienvenido al Ordenamientos en Java-------------------");
        do {
            System.out.println("\n \nMenú para ejecutar los algoritmos de ordenamiento a nuestro Array: edades_java");
            System.out.println("1. Insertion Sort (Ordenamiento por inserción)");
            System.out.println("2. Bubble Sort (Ordenamiento de burbuja)");
            System.out.println("3. Selection Sort (Ordenamiento por selección)");
            System.out.println("4. Merge Sort (Ordenamiento por mezcla)");
            System.out.println("5. Random Sort (Ordenamiento aleatorio)");
            System.out.println("6. Salir");
            System.out.println("Ingresa un número del 1 al 5 para ejecutar un algoritmo de ordenamiento (o 6 para salir): ");

            String opcion = scanner.nextLine();
            /* usamos un switch para evaluar que tecla se introdujo con el teclado 
            de esta manera llamamos a las funcion correspondiente
            a cada funcion le pasamos como parametros el array edades_java 
            el cual contine los elementos que usara cada funcion*/
            switch (opcion) {
                case "1":
                    // Antes de ordenar
                    System.out.println("\nAntes de ordenar:");
                    for (int i = 0; i < edades_java.length; i++) {
                        System.out.print(edades_java[i] + " ");
                    }
                    // Llamada al método de ordenamiento
                    O_Insercion(edades_java);

                    // Después de ordenar
                    System.out.println("\nDespués de ordenar con O_Insercion:");
                    for (int edad : edades_java) {
                        System.out.print(edad + " ");
                    }
                    break;
                case "2":
                    // Antes de ordenar
                    System.out.println("\nAntes de ordenar:");
                    for (int i = 0; i < edades_java.length; i++) {
                        System.out.print(edades_java[i] + " ");
                    }
                    // Llamada al método de ordenamiento
                    O_Burbuja(edades_java);

                    // Después de ordenar
                    System.out.println("\nDespués de ordenar con O_Burbuja:");
                    for (int edad : edades_java) {
                        System.out.print(edad + " ");
                    }
                    break;
                case "3":
                    // Antes de ordenar
                    System.out.println("\nAntes de ordenar:");
                    for (int i = 0; i < edades_java.length; i++) {
                        System.out.print(edades_java[i] + " ");
                    }
                    // Llamada al método de ordenamiento
                    O_Select(edades_java);

                    // Después de ordenar
                    System.out.println("\nDespués de ordenar con O_Select:");
                    for (int edad : edades_java) {
                        System.out.print(edad + " ");
                    }
                    break;
                case "4":// Antes de ordenar
                    System.out.println("\nAntes de ordenar:");
                    for (int i = 0; i < edades_java.length; i++) {
                        System.out.print(edades_java[i] + " ");
                    }
                    // Llamada al método de ordenamiento
                    O_Mezcla(edades_java, 0, edades_java.length - 1);

                    // Después de ordenar
                    System.out.println("\nDespués de ordenar con O_Mezcla:");
                    for (int edad : edades_java) {
                        System.out.print(edad + " ");
                    }
                    break;
                case "5":
                    // Antes de ordenar
                    System.out.println("\nAntes de ordenar:");
                    for (int i = 0; i < edades_java.length; i++) {
                        System.out.print(edades_java[i] + " ");
                    }
                    // Llamada al método de ordenamiento
                    O_Random(edades_java);

                    // Después de ordenar
                    System.out.println("\nDespués de ordenar con O_Select:");
                    for (int edad : edades_java) {
                        System.out.print(edad + " ");
                    }
                    break;
                case "6":
                    System.out.println("Saliendo del programa...");
                    return;
                default:
                    //cualqueir otro digito pro fuera de los elegidos incialmente se notifica quee sta incorrecto
                    System.out.println("Opción inválida. Por favor, ingresa un número del 1 al 6.");
            }
        } while (true);
    }

    public static void O_Insercion(int[] edadesi) {
        int n = edadesi.length;

        for (int i = 1; i < n; i++) {
            int valorActual = edadesi[i];
            int j = i - 1;

            // Desplazar los elementos mayores que el valor actual hacia la derecha
            while (j >= 0 && edadesi[j] > valorActual) {
                edadesi[j + 1] = edadesi[j];
                j = j - 1;
            }

            // Insertar el valor actual en la posición correcta
            edadesi[j + 1] = valorActual;
        }
    }

    public static void O_Burbuja(int[] edadesb) {
        int n = edadesb.length;

        // Iteramos sobre el arreglo n veces
        for (int i = 0; i < n - 1; i++) {
            // En cada iteración, comparamos pares de elementos que esatan al lado del elemento a comparar
            for (int j = 0; j < n - i - 1; j++) {
                // Si el elemento actual es mayor que el siguiente, los intercambiamos es decir derecha o izquierda
                if (edadesb[j] > edadesb[j + 1]) {
                    int temp = edadesb[j];
                    edadesb[j] = edadesb[j + 1];
                    edadesb[j + 1] = temp;
                }
            }
        }
    }

    public static void O_Select(int[] edadess) {
        int n = edadess.length;

        // Iteramos sobre el arreglo
        for (int i = 0; i < n - 1; i++) {
            // Encontramos el índice del mínimo elemento en el subarreglo no ordenado
            int minIndex = i;
            for (int j = i + 1; j < n; j++) {
                if (edadess[j] < edadess[minIndex]) {
                    minIndex = j;
                }
            }

            // Intercambiamos el mínimo elemento con el primer elemento del subarreglo no ordenado
            int temp = edadess[minIndex];
            edadess[minIndex] = edadess[i];
            edadess[i] = temp;
        }
    }

    public static void O_Mezcla(int[] edadesm, int izquierda, int derecha) {
        if (izquierda < derecha) {
            // Encuentra el punto medio del arreglo
            int medio = (izquierda + derecha) / 2;

            // Ordena la primera mitad
            O_Mezcla(edadesm, izquierda, medio);
            // Ordena la segunda mitad
            O_Mezcla(edadesm, medio + 1, derecha);

            // Une las mitades ordenadas
            Mezcla(edadesm, izquierda, medio, derecha);
        }
    }

    //funcion adicional para mezclar
    public static void Mezcla(int[] edadesm, int izquierda, int medio, int derecha) {
        // Tamaño de los subarreglos
        int n1 = medio - izquierda + 1;
        int n2 = derecha - medio;

        // Arreglos temporales
        int[] izq = new int[n1];
        int[] der = new int[n2];

        // Copiar datos a los arreglos temporales
        for (int i = 0; i < n1; i++) {
            izq[i] = edadesm[izquierda + i];
        }
        for (int j = 0; j < n2; j++) {
            der[j] = edadesm[medio + 1 + j];
        }

        // Índices iniciales de los subarreglos
        int i = 0, j = 0;
        // Índice inicial del arreglo fusionado
        int k = izquierda;

        // Fusionar los arreglos temporales
        while (i < n1 && j < n2) {
            if (izq[i] <= der[j]) {
                edadesm[k] = izq[i];
                i++;
            } else {
                edadesm[k] = der[j];
                j++;
            }
            k++;
        }

        // Copiar elementos restantes de izq[] si los hay
        while (i < n1) {
            edadesm[k] = izq[i];
            i++;
            k++;
        }

        // Copiar elementos restantes de der[] si los hay
        while (j < n2) {
            edadesm[k] = der[j];
            j++;
            k++;
        }
    }

    public static void O_Random(int[] edadesr) {
        Random rand = new Random();
        int n = edadesr.length;

        // Realiza un número grande de iteraciones (n^2) para aumentar las posibilidades de que se ordene
        for (int i = 0; i < n * n; i++) {
            // Selecciona dos índices aleatorios
            int index1 = rand.nextInt(n);
            int index2 = rand.nextInt(n);

            // Intercambia los elementos en los índices seleccionados
            int temp = edadesr[index1];
            edadesr[index1] = edadesr[index2];
            edadesr[index2] = temp;
        }
    }

}
